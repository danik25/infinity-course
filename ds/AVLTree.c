/****************************************************************
*                       AVL Binary Search Tree
*                       =====================
* File Name: AVLTree.h
* Related files: AVLTree.c, AVLTree_test.c
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Moshe Greenberg    
* Last update: 02/06/19
****************************************************************/
#include <assert.h> /*assert */
#include <stdlib.h> /* malloc, abs */

#include "AVLTree.h"

enum CHILDREN {LEFT, RIGHT};
#define MAX(x, y) (((x) > (y)) ? (x) : (y));

typedef struct node
{
    void *data;
    struct node *children[2];
    size_t height;
}AVL_node_t;

struct AVL
{
    AVL_node_t *root;
    compare_t CompareFunc;
    void *param;
};

/* lut for rotation cases, used in balance function */
typedef AVL_node_t *(*rotation_case_func)
                    (AVL_node_t *root, AVL_node_t *pivot, int direction);


/********************** assistance functions for remove **********************/

static AVL_node_t* RecFindParent(AVL_node_t *node, compare_t CompareFunc, 
                                        const void *data, void* param);
static AVL_node_t* RecFindNodeToRemove(AVL_node_t *node, compare_t CompareFunc, 
                                        const void *data, void* param);
static void RecRemoveOneChildNode(AVL_node_t *parent_node, 
                                        void *data, AVL_t *tree);
static void RecRemoveRoot(AVL_node_t *root, AVL_t *tree);
static AVL_node_t* FindTheNextNode(AVL_node_t *node);
static void RecHeightUpdate(AVL_node_t *root, AVL_node_t *removed_parent, 
                                        compare_t CompareFunc, void *param);
static AVL_node_t* ComplicatedCase(AVL_node_t *root, AVL_node_t *pivot, 
                                                        int rot_direction);

/******************************************************************************/


/************************ general assistance functions ************************/

static int HasOneChild(AVL_node_t *node);
static int ChildDirection(AVL_node_t *node);
static int Isleaf(AVL_node_t *node);
static int HasTwoChildren(AVL_node_t *node);
static AVL_node_t* NodeCreate(void *data);

/*****************************************************************************/


/********************** recursive assistance functions **********************/

static AVL_node_t* RecDestroy(AVL_node_t *node);
static int RecInsert(AVL_node_t *tree_node, AVL_node_t *new_node, AVL_t *tree);

static AVL_node_t* RecFind(AVL_node_t *node, compare_t CompareFunc, 
                                    const void *data, void* param);
static int RecCount(AVL_node_t *node);
static int RecForEach(AVL_node_t *node, action_func_t ActionFunc, void* param);

/*****************************************************************************/


/****************************** balancing functions ***************************/

static AVL_node_t* Rotate(AVL_node_t *root, AVL_node_t *pivot, int direction);
static int HeightGap(AVL_node_t *node);
static void Balance(AVL_t *tree, AVL_node_t *sub_root, int balance_factor);
static void HeightUpdate(AVL_node_t *node);
static size_t GetNodeHeight(AVL_node_t *node);

/*****************************************************************************/

/******************************************************************************

                        ************* API *****************

******************************************************************************/

AVL_t *AVLCreate(compare_t CompareFunc, void *param)
{
    AVL_t * tree_handle = NULL;

    assert(CompareFunc);

    tree_handle = (AVL_t*)malloc(sizeof(AVL_t));

    if(NULL == tree_handle)
    {
        return (NULL);
    }

    tree_handle->CompareFunc = CompareFunc;
    tree_handle->param = param;
    tree_handle->root = NULL;

    return(tree_handle);
}

void AVLDestroy(AVL_t *tree)
{
    assert(tree);

    RecDestroy(tree->root);

    free(tree);
    tree = NULL;   
}

int AVLInsert(AVL_t *tree, void *data)
{
    AVL_node_t * new_node = NULL;

    assert(tree);

    new_node = NodeCreate(data);

    /* failure in creating new node */
    if(!new_node)
    {
        return (1);
    }

    if(AVLIsEmpty(tree))
    {
        tree->root = new_node;
    }
    else
    {
        return(RecInsert(tree->root, new_node, tree));
    }

    return (0);    
}

void AVLRemove(AVL_t *tree, void *data)
{
    AVL_node_t *node = NULL;
    AVL_node_t *next_node = NULL;
    AVL_node_t * remove_parent = NULL;
    void *data_to_replace = NULL;
    int balance_factor = 0;

    assert(tree);

    /* if the data inserted for removal doesnt exists */
    if(!AVLFind(tree, data))
    {
        return ;
    }

    node = RecFindNodeToRemove(tree->root, tree->CompareFunc, data, 
                                                        tree->param);
    /* has two children */
    if(HasTwoChildren(node))
    {
        /* finding the next node and saving it's data */
        next_node = FindTheNextNode(node->children[RIGHT]);
        data_to_replace = next_node->data;

        /* finding the parent of the "next node", 
                                the next node needs to be removed now */
        remove_parent = RecFindParent(tree->root, 
                            tree->CompareFunc, next_node->data, tree->param);

        /* removing the node that it's data was copied */
        RecRemoveOneChildNode(remove_parent, next_node->data, tree);

        node->data = data_to_replace;    

        RecHeightUpdate(tree->root, remove_parent, tree->CompareFunc, 
                                                            tree->param);
        return ;
    }

    /* if the node to be removed is the root */
    if(0 == tree->CompareFunc(tree->root->data, data, tree->param))
    {
        if(Isleaf(node))
        {
            free(node);
            node = NULL;
            tree->root = NULL;
            return ;
        }
        RecRemoveRoot(tree->root, tree);
        return ;
    }

    /* in case the node to be removed has an only child or no children at all */
    if(Isleaf(node) || HasOneChild(node))
    {
        remove_parent = RecFindParent(tree->root, 
                        tree->CompareFunc, data, tree->param);
        RecRemoveOneChildNode(remove_parent, data, tree);
    }

    balance_factor = HeightGap(tree->root);
    if(abs(balance_factor) > 1)
    {
        Balance(tree, tree->root, balance_factor);
        HeightUpdate(tree->root);
    }
}


void *AVLFind(const AVL_t *tree, const void *data)
{
    assert(tree);

    if(!AVLIsEmpty(tree))
    {
        return(RecFind(tree->root, tree->CompareFunc, data, tree->param));
    }
    
    return(NULL);
}


size_t AVLCount(const AVL_t *tree)
{
    assert(tree);

    if(AVLIsEmpty(tree))
    {
        return(0);
    }
    return(RecCount(tree->root));
}

int AVLIsEmpty(const AVL_t *tree)
{
    assert(tree);

    return(NULL == tree->root);
}

size_t AVLHeight(const AVL_t *tree)
{
    assert(tree);

    if(AVLIsEmpty(tree))
    {
        return(0);
    }

    return(tree->root->height);
}

int AVLForEach(AVL_t *tree, action_func_t ActionFunc, void *param)
{
    assert(tree);
    assert(ActionFunc);

    if(AVLIsEmpty(tree))
    {
        return(1);
    }

    return(RecForEach(tree->root, ActionFunc, param));
}



/******************************************************************************

            ************* balancing functions *****************

******************************************************************************/

static AVL_node_t* Rotate(AVL_node_t *root, AVL_node_t *pivot, int direction)
{
    assert(root);
    assert(pivot);

    root->children[direction] = pivot->children[!direction];
    pivot->children[!direction] = root;

    HeightUpdate(root);
    HeightUpdate(pivot);

    return(pivot);
}

static int HeightGap(AVL_node_t *node)
{
    /* the difference between the height of the 
                        right node and the left node is greater than 1 */
    return(GetNodeHeight(node->children[RIGHT]) - 
                        GetNodeHeight(node->children[LEFT]));
}

static AVL_node_t* ComplicatedCase(AVL_node_t *root, AVL_node_t *pivot, 
                                                        int rot_direction)
{
    /* left rotation */
    pivot = Rotate(root->children[rot_direction], 
                root->children[rot_direction]->children[!rot_direction], 
                                            !rot_direction);
    root->children[rot_direction] = pivot;

    pivot = Rotate(root, pivot, rot_direction);

    HeightUpdate(root);
    HeightUpdate(pivot);

    return (pivot);
}

static void Balance(AVL_t *tree, AVL_node_t *sub_root, int balance_factor)
{
    AVL_node_t *pivot = NULL;
    AVL_node_t *root_parent = NULL;
    int direction = 0;
    int rotate_direction = 0;
    int height_gap = 0;

    /*
        _|_______|_______0__________|__________1___________|  
         |       |  height_gap < 0  |   height_gap > 0     |
      ___|_______|__________________|______________________|
       0 |gap < 0| Rotate           | ComplicatedCase      |
         |       | (left left)      | (left right)         |
      ___|_______|__________________|______________________|
       1 |gap > 0| ComplicatedCase  | Rotate               |
         |       | (right left)     | (right right)        |
      ___|_______|__________________|______________________|
    
    */

    rotation_case_func rot_lut[2][2] = 
    {
        {Rotate, ComplicatedCase},
        {ComplicatedCase, Rotate}
    };

    rotate_direction = (balance_factor > 0);
    height_gap = (HeightGap(sub_root->children[rotate_direction]) > 0);

    pivot = (*rot_lut[rotate_direction][height_gap])
            (sub_root, sub_root->children[rotate_direction], rotate_direction);
 

    /* if the balancing is happening on the root */
    if(sub_root->data == tree->root->data)
    {
        tree->root = pivot;
    }

    else
    {
        root_parent = RecFindParent(tree->root, 
                        tree->CompareFunc, sub_root->data, tree->param);
        direction = tree->CompareFunc(sub_root->data, 
                            root_parent->data, tree->param) > 0 ? RIGHT : LEFT;
        root_parent->children[direction] = pivot;
    }

    HeightUpdate(pivot); 
}

static void HeightUpdate(AVL_node_t *node)
{
    node->height = 1 + MAX(GetNodeHeight(node->children[LEFT]), 
                GetNodeHeight(node->children[RIGHT]));
}
/******************************************************************************

            ************* general assistance functions *****************

******************************************************************************/

static AVL_node_t* NodeCreate(void *data)
{
    AVL_node_t *node = (AVL_node_t*)malloc(sizeof(AVL_node_t));
    if(!node)
    {
        return (NULL);
    }
    node->data = data;
    node->height = 1;
    node->children[LEFT] = NULL;
    node->children[RIGHT] = NULL;

    return(node);
}

static int HasOneChild(AVL_node_t *node)
{
    return((NULL != node->children[LEFT] && NULL == node->children[RIGHT]) ||
           (NULL == node->children[LEFT] && NULL != node->children[RIGHT]));
}

static int ChildDirection(AVL_node_t *node)
{
    /* if the left child is NULL, returns 1, which is "right". and vise versa */
    return(NULL == node->children[LEFT]);
}

static int Isleaf(AVL_node_t *node)
{
    return(NULL == node->children[LEFT] && NULL == node->children[RIGHT]);
}

static int HasTwoChildren(AVL_node_t *node)
{
    return(NULL != node->children[LEFT] && NULL != node->children[RIGHT]);
}

static size_t GetNodeHeight(AVL_node_t *node)
{
    if(NULL == node)
    {
        return (0);
    }
    return(node->height);
}

/******************************************************************************

              ************* static recursive functions *****************

******************************************************************************/

static int RecInsert(AVL_node_t *tree_node, 
                    AVL_node_t *new_node, AVL_t *tree)
{
    int direction = tree->CompareFunc(new_node->data, tree_node->data, 
                                                                tree->param); 
    int ans = 0;  
    int balance_factor = 0;              
    if(0 == direction)
    {
        return (1);
    }

    direction = (direction > 0) ? RIGHT : LEFT;

    if(NULL == tree_node->children[direction])
    {
        tree_node->children[direction] = new_node;
        if(NULL == tree_node->children[!direction])
        {
            tree_node->height += 1;
        }
        return(0) ;
    }
    ans = RecInsert(tree_node->children[direction], new_node, 
                                                    tree);
    HeightUpdate(tree_node);

    balance_factor = HeightGap(tree_node);
    if(abs(balance_factor) > 1)
    {
        Balance(tree, tree_node, balance_factor);
    }

    return(ans);

}

static AVL_node_t* RecFind(AVL_node_t *node, compare_t CompareFunc, 
                                                const void *data, void* param)
{
    int direction = CompareFunc(data, node->data, param);
                                                                       
    if(direction == 0)
    {
        return(node->data);
    }
    direction = (direction > 0);

    if(NULL == node->children[direction])
    {
        return (NULL);
    }

    return(RecFind(node->children[direction], CompareFunc, data, param));
}

static int RecForEach(AVL_node_t *node, action_func_t ActionFunc, void* param)
{
    if(NULL == node)
    {
        return(0);
    }
    
    if(RecForEach(node->children[LEFT], ActionFunc, param))
    {
        return(1);
    }
    if(ActionFunc(node->data, param))
    {
        return(1);
    }
    if(RecForEach(node->children[RIGHT], ActionFunc, param))
    {
        return(1);
    }
    return(0);
}

static AVL_node_t * RecDestroy(AVL_node_t *node)
{
    if(NULL != node)
    {
        RecDestroy(node->children[LEFT]);    
        RecDestroy(node->children[RIGHT]);

        free(node);
    }

    return(node);   
}

static int RecCount(AVL_node_t *node)
{
    if(NULL == node)
    {
        return(0);
    }
    
    return (1 + RecCount(node->children[LEFT]) + 
                                    RecCount(node->children[RIGHT]));
}

/*****************************************************************************/

                        /* remove assistance functions */

/*****************************************************************************/

static AVL_node_t* RecFindParent(AVL_node_t *node, 
                        compare_t CompareFunc, const void *data, void* param)
{
    int direction = (CompareFunc(data, node->data, param) > 0) ? RIGHT : LEFT;
                                                                       
    if(0 == CompareFunc(data, node->children[direction]->data, param))
    {
        return(node);
    }

    return(RecFindParent(node->children[direction], 
                                            CompareFunc, data, param));
}

static AVL_node_t* RecFindNodeToRemove(AVL_node_t *node, compare_t CompareFunc, 
                                                const void *data, void* param)
{
    int direction = (CompareFunc(data, node->data, param) > 0) ? RIGHT : LEFT;
                                                                       
    if(0 == CompareFunc(data, node->data, param))
    {
        return(node);
    }

    return(RecFindNodeToRemove(node->children[direction], CompareFunc, 
                                                            data, param));
}

static void RecRemoveOneChildNode(AVL_node_t *parent_node, 
                                                    void *data, AVL_t *tree)
{
    int gap = 0;
    /* the direction of the node to be removed from the parent */
    int remove_direction = 
        (tree->CompareFunc(data, parent_node->data, tree->param)) > 0 ? 
                                                                RIGHT : LEFT;

    /* the direction of the child of the node to be removed */
    int child_direction = 
                    ChildDirection(parent_node->children[remove_direction]); 

    /* the child of the node to be removed */
    AVL_node_t *new_child = 
            parent_node->children[remove_direction]->children[child_direction];

    free(parent_node->children[remove_direction]);
    parent_node->children[remove_direction] = new_child;
    
    HeightUpdate(parent_node);

    RecHeightUpdate(tree->root, parent_node, tree->CompareFunc, tree->param);

    gap = HeightGap(parent_node);
    if(abs(gap) > 1)
    {
        Balance(tree, parent_node, gap);
        HeightUpdate(parent_node);
    }
    
}

static void RecRemoveRoot(AVL_node_t *root, AVL_t *tree)
{
    /* if the root has only one child */
    if(HasOneChild(root))
    {
        tree->root = root->children[ChildDirection(root)];
        free(root);
    }
}

static AVL_node_t* FindTheNextNode(AVL_node_t *node)
{
    AVL_node_t *ret_node = NULL;

    if(NULL != node->children[LEFT])
    {
        ret_node = FindTheNextNode(node->children[LEFT]); 
        return(ret_node);
    }
    return(node);
}

static void RecHeightUpdate(AVL_node_t *root, AVL_node_t *removed_parent, 
                                        compare_t CompareFunc, void *param)
{
    int direction = CompareFunc(removed_parent->data, root->data, param) > 0 
                                                            ? RIGHT : LEFT;
    if(root->data == removed_parent->data)
    {
        return ;
    }

    RecHeightUpdate(root->children[direction], removed_parent, CompareFunc, 
                                                                    param);

    HeightUpdate(root);
}
