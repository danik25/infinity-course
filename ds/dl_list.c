/****************************************************************
*                Data Structures - Double Linked List           *
* Related files: dl_list.c 	dl_list.h	dl_list_test.c	        *
* #Version: V 1.7                                               *
* writer: Dani Benjamin                                 		*
* reviewer: Amir Tzin                               			*
* last update: 14.4.19                                          *
****************************************************************/
#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/
#include "dl_list.h"

#define DEADBEEF (void*)0xdeadbeefdeadbeef

typedef struct dl_node {

	void *data;
	struct dl_node *next;
	struct dl_node *prev;

}dl_node_t;

struct dl_list {

	dl_node_t head;
	dl_node_t tail;
};

static dl_node_t *DLLNodeCreate(void *data, dl_node_t *next, dl_node_t *prev);

/******************************************************************************/

/*assistance function!!*/
static dl_node_t *DLLNodeCreate(void *data, dl_node_t *next, dl_node_t *prev)
{
    dl_node_t *node = NULL;
    
    node = (dl_node_t*) malloc(sizeof(dl_node_t));
    
    if (NULL == node)
    {
        return (NULL);
    }
    node->data = data;
    node->next = next;
    node->prev = prev;
    
    return (node); 
}

/******************************************************************************/

dll_t *DLCreate()
{
	dll_t *dll = (dll_t*) malloc(sizeof(dll_t));
    if(!dll)
    {
        return (NULL);
    }
    dll->head.data = DEADBEEF;
    dll->tail.data = DEADBEEF;
    
    dll->head.next = &(dll->tail);
    dll->head.prev = NULL;
    
    dll->tail.next = NULL;
    dll->tail.prev = &(dll->head);
    
    return (dll);
}

void DLDestroy(dll_t *dl_list)
{
	dl_node_t *node = NULL;
	
	assert(dl_list);
	
	node = DLBegin(dl_list);

    while(node != DLEnd(dl_list))
    {
        node = DLErase(node);
    }
	
	free(dl_list);
    dl_list = NULL;
}

int DLIsEmpty(const dll_t *dl_list)
{
    assert(dl_list);
	return (DLSameIter(DLBegin(dl_list), DLEnd(dl_list)));
}

size_t DLCount(const dll_t *dl_list)
{
	size_t cnt = 0;
	dl_node_t *node = NULL;
	
	assert(dl_list);

	for(node = DLBegin(dl_list); node != DLEnd(dl_list); node = DLNext(node))
	{
		++cnt;
	}
	return (cnt);
}

dl_iter_t DLEnd(const dll_t *dl_list)
{
    assert(dl_list);
	return (dl_iter_t)&(dl_list->tail);
}

dl_iter_t DLBegin(const dll_t *dl_list)
{
    assert(dl_list);
	return (dl_list->head.next);
}

dl_iter_t DLNext(const dl_iter_t location)
{
    assert(location);

	return (location->next);
}

dl_iter_t DLPrev(const dl_iter_t location)
{
    assert(location);

	return (location->prev);
}

void *DLGetData(const dl_iter_t location)
{
    assert(location);
	return (location->data);
}

int DLSameIter(const dl_iter_t first_iter, const dl_iter_t second_iter)
{
    assert(first_iter);
    assert(second_iter);
	return(first_iter == second_iter);
}

dl_iter_t DLInsert(dll_t *dl_list, dl_iter_t location, void *data_to_insert)
{
	dl_node_t *new_node = NULL;
	
	assert(dl_list);
	assert(location);
    
    new_node = DLLNodeCreate(data_to_insert, NULL, NULL);
	if(!new_node)
	{
		return (&dl_list->tail);
	}
	
	new_node->prev = location->prev;
	new_node->next = location;
	
	location->prev->next = new_node;
	location->prev = new_node;
	
	return 	(new_node);
}

dl_iter_t DLErase(dl_iter_t location)
{
    dl_node_t *next_node = NULL;
    
    assert(location);
    
    if(NULL == location->next || NULL == location->prev)
    {
        return (NULL);
    }
    next_node = location->next;
    
    location->prev->next = location->next;
    location->next->prev = location->prev;
    
    free(location);
    location = NULL;
    
    return (next_node);
}

dl_iter_t DLPushFront(dll_t *dl_list, void *data)
{
	assert(dl_list);
	
	return (DLInsert(dl_list, DLBegin(dl_list), data));
}

dl_iter_t DLPushBack(dll_t *dl_list, void *data)
{
    assert(dl_list);
	
	return (DLInsert(dl_list, DLEnd(dl_list), data));
}

void DLPopFront(dll_t *dl_list)
{
    assert(dl_list);
    DLErase(DLBegin(dl_list));
}

void DLPopBack(dll_t *dl_list)
{    
    assert(dl_list);
    DLErase(DLEnd(dl_list)->prev);
}

int DLForEach(dl_iter_t from, dl_iter_t to,
						 	action_function_t func , void *param)
{
    dl_iter_t node = from;
    int status = 0;
    
    assert(func);
    assert(to);
    assert(from);
    for(; node != to && (!status); node = DLNext(node)) 
    {
        status = func(node->data, param);
    }
    return (status);
}

dl_iter_t DLFind(dll_t *dl_list, dl_iter_t from, dl_iter_t to,
 		compare_function_t func, void *param)
{
    dl_iter_t node = NULL;
    
    assert(dl_list);
    assert(func);
    assert(from);
    assert(to);
    
    node = from;
    while(node != to && (!func(node->data, param)))
    {
        node = DLNext(node);
    }
    return (node);
}

void DLSplice(dll_t *src_list, dl_iter_t from, dl_iter_t to, dl_iter_t location)
{
    dl_iter_t first_src = NULL;
    dl_iter_t last_src = NULL; 
    
    assert(src_list);
    assert(from);
    assert(to);
    assert(location);
    
    if(DLSameIter(to, from))
    {
        return;
    }
    
    /*attaching the from remains*/
    from->prev->next = to;
    
    first_src = from;
    last_src = to;
    
    last_src = last_src->prev;
    
    /*attaching the to remains*/
    to->prev = from->prev;
    
    location->prev->next = first_src;
    first_src->prev = location->prev;
    location->prev = last_src;
    last_src->next = location; 
}


