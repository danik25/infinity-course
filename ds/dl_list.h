/****************************************************************
*                Data Structures - Double Linked List           *
* Related files: dl_list.c 	dl_list.h	dl_list_test.c	        *
* #Version: V 1.7                                               *
* writer: Dani Benjamin                                 		*
* reviewer: Amir Tzin                               			*
* last update: 14.4.19                                          *
****************************************************************/
/*
* A Double Linked List contains an extra pointer, typically called prev 
* pointer, together with next pointer and data which are there in singly linked 
* list. The main advantage of double linked lists over single linked list 
* is A dl_list can be traversed in both forward and backward direction.
* We can quickly insert a new node before a given node. As a convention, the 
* first element we referd to is the one next to the head dummy, and the last
* element is the one previous to tail dummy.
*/  

/* API Design:
 *	
 *		This API design is an infrastructure module, allowing the user
 *		to access the structure.
 *		With the ability to access the structure, the user can add nodes,
 *		delete nodes, modify the structure itself as well as the data in
 *		any of it's fields.
 *
 *		This design supports any data type. 
 * 	                

	  		 	****************** <-- *****************<-- *****************    
NULL<--DUMMY <--* DATA  *   next * --> *  DATA  * next * -->*  DATA  * next *---> DUMMY-->NULL 
	   			******************     *****************    *****************
	   		  	    1st node 		    2nd node  ....n'th node...    Last node	
*/
#ifndef _DL_LIST_
#define _DL_LIST_

#include <stddef.h> /* size_t */

/* typedef on function pointer for readability - both functions are given */
/* as arguments to DLLForEach and DLLFind respectively. 
success(0) or failure(1).*/

typedef int (action_function_t)(void *data, void* param);
typedef int (compare_function_t)(void *data, void* param);

typedef struct dl_node* dl_iter_t;
typedef struct dl_list dll_t;


/* Create an empty doubly linked list, actually containing two dummy nodes.
 * Params: void.
 *		  
 * Return Value: 
 *		  pointer to the newly created list. 
*/
dll_t *DLCreate();


/* Destroy the doubly linked list - free all allocated memory from a given position.
 *
 * param: dl_list - the list to work on.
 * Time complexity: O(n).
*/
void DLDestroy(dll_t *dl_list);	  


/* Checks whether or not the list is empty.
 * 
 * param: dl_list - the list to work on.
 *
 * Return Value: boolean (TRUE(1) / FALSE(0)).
 *
*/
int DLIsEmpty(const dll_t *dl_list);


/* Counts elements in a given list.
 *  
 * param: dl_list - the list to work on.
 *
 * Return value: number of elements in the list.
*/
size_t DLCount(const dll_t *dl_list);


/* Finds the last element of a given list. 
 * 
 * param: dl_list - the list to work on.	
 *	  
 * Return value: a pointer to the end of the list.
*/
dl_iter_t DLEnd(const dll_t *dl_list);


/* Finds the first element of a given list. 
 *	 
 * param: dl_list - the list to work on.
 *	  
 * Return value: a pointer to the beginning of the list.
 * in case of empty list return dummy tail.
*/
dl_iter_t DLBegin(const dll_t *dl_list);


/* Gives the next element from a given location. 
 *
 * param: location - specific node to check.
 *		  
 * Return value: a pointer to the next element of a given location. 
 * If location is DLEnd()- undefined behavior. 
*/
dl_iter_t DLNext(const dl_iter_t location);


/* Gives the previous element from a given location. 
 *	
 * param: 
 *           location -   specific node to check.
 *                 
 * Return value: a pointer to the prev element of a given location.
 *  If location is DLBegin()- undefined behavior. 
*/
dl_iter_t DLPrev (const dl_iter_t location);


/* Extracts data from a given element.
 *
 * param: location - specific node to check.
 *
 * Return value: the data from the given location. 
*/
void *DLGetData(const dl_iter_t location);


/* Compares addresses of two given iterators.
 * position.
 *
 * Params: first_iter - first pointer to a node. 
 *		   second_iter - second pointer to a node.
 *
 * Return Value: boolean (TRUE(1) / FALSE(0)).
*/
int DLSameIter(const dl_iter_t first_iter, const dl_iter_t second_iter);


/* Creates and inserts an element before a given location.
 * Params:
 *		  dl_list - A list to insert an element to.
 *		  location - a reference location.		 
 *		  data_to_insert - the data to insert.	
 * Return value : 
 *        pointer to inserted node if succeded. tail dummy if failed.  
*/
dl_iter_t DLInsert(dll_t *dl_list, dl_iter_t location, void *data_to_insert);


/* Removes an element from a list and erases it's content.
 * Params: 
 *		  location - the element to be removed.
 *		  
 * Return value : 
 *        pointer to the element after the removed one if succeded.
 *		  If location is DLBegin()- undefined behavior. 
*/
dl_iter_t DLErase(dl_iter_t location);


/* Create and insert an element to the beginning of the list.
 * Params: 
 *		  dl_list - A list to insert an element to.
 *		  data - data of the inserted element.
 * Return value : 
 *        pointer to inserted element or end dummy if fails. 
*/
dl_iter_t DLPushFront(dll_t *dl_list, void *data);


/* Create and insert an element to the end of the list.
 * Params: 
 *		  dl_list - a list to insert an element to.
 *		  data - data of the inserted element.
 * Return value : 
 *        pointer to inserted element or end dummy if fails. 
*/
dl_iter_t DLPushBack(dll_t *dl_list, void *data);


/* Remove a node from the beginning of the list.
 * Params: 
 *		  dl_list - a list to remove an element from.
*/
void DLPopFront(dll_t *dl_list);


/* Remove a node from the end of the list.
 * Params: 
 *		  dl_list - a list to remove an element from.
*/
void DLPopBack(dll_t *dl_list);


/* Apply a provided function on a given sequence of elements. 
 * Params: 
 *		  from - start position.
 * 		  to - end position (up to and not including).
 *		  func - user provided function.
 * 		  param - additional parameter provided by user.
 * Return value : 
 *        status - success(0) or failure(1). 
*/
int DLForEach(dl_iter_t from, dl_iter_t to,
						 	action_function_t func , void *param);


/* Search if a given data exists on a provided list. 
 * Params: 
 *		  dl_list - a list from which the sequence is given.
 *		  from - start position.
 * 		  to - end position (up to and not including).
 *        param - data to find.
 * Return value : 
 *        pointer to the match or end dummy if no match.
*/
dl_iter_t DLFind(dll_t *dl_list, dl_iter_t from, dl_iter_t to,
 		compare_function_t func, void *param);


/* Insert a sequence of elements before a given location.
 * Params: 
 *		  src_list - the list to transfer the sequence from. 
 *		  from - start position.
 * 		  to - end position (up to and not including).
 *		  location - node to insert the sequence before.
 *        in case src_list is the same list as location list the behaiver is 
 *        undefined.
*/
void DLSplice(dll_t *src_list, dl_iter_t from, dl_iter_t to, dl_iter_t location);

#endif /* _DL_LIST_ */

