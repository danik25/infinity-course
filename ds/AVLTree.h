#ifndef  _AVL_H_
#define  _AVL_H_

#include <stddef.h> /* size_t */

#include "dl_list.h"

/****************************************************************
*                       AVL Binary Search Tree
*                       =====================
* File Name: AVLTree.h
* Related files: AVLTree.c, AVLTree_test.c
* #Version: V 1.1
* Writer: Dani Benjamin      
* Reviewer: Moshe Greenberg     
* Last update: 02/06/19
****************************************************************/

/* API Design:
 * An AVL Binary Search Tree, Following the same restrictions as a BST
 * With an added feature of self balancing.
 * this structures is guarenteed to be always balanced as per conventional 
 * definition of a balanced tree structure.
 *
 * This AVL Tree DOES NOT support duplicate data.
 * In our structure the Height of an empty tree, as will be returned from
 * AVLHeight, will be 0. Height of a leaf will be 1.
 * 
 */

/*** Services and Auxiliaries: ***/

/* compare_t  DESCRIPTION:
 *
 * return values: an integer "num" such that:
 *
 *  num < 0 if data1 < data2
 *  num = 0 if data1 = data2
 *  num > 0 if data1 > data2
*/

typedef int (*compare_t)(const void *data1, const void *data2, void *param);
typedef int (*is_match_t)(void *data, void *param);

 /* action function DESCRIPTION:
 *
 * a function that perform user action on each data.
 * return val : 0 = success, otherwise failure according to it's failure code.
 */
typedef int (*action_func_t)(void *data, void *param);

typedef struct AVL AVL_t;


/* Description:
 * Creates a new binary search tree .
 *
 * Params: CompareFunc - A pointer to a function to determine the order in which
 *                     key are sorted in the structure.
 *        param - parameter for Compare function.
 *
 * Return Value: A pointer to the tree data structure.
 *               NULL if failed.
 */
AVL_t *AVLCreate(compare_t CompareFunc, void *param);


/* Description: destroys the given tree .
 *
 * Params: tree - A pointer to the tree data structure.
 *
*/
void AVLDestroy(AVL_t *tree);


/* Description: insert new data to the tree.
 * 
 * Params: tree - A pointer to the tree data structure.
 *         data - data to insert.
 *
 * Return Value: Success (0) or:
 *               1 -  memeory allocation failure;
 *               2 -  element duplication failure;
 * 
*/
int AVLInsert(AVL_t *tree, void *data);


/* Description: remove data from the tree. 
 *
 * Params: tree - the tree from which the data is to be removed.
 *         data - data to be removed.
*/
void AVLRemove(AVL_t *tree, void *data);


/* Description: find specific data in the tree.
 *
 * Params: tree - A pointer to the tree in which to find data.
 *         data - A pointer to data to find.
 *
 * Return Value: pointer to data supplied by the user.
 *               NULL if data not found.
 *
 * Time complexity: O(log(n)).
*/
void *AVLFind(const AVL_t *tree, const void *data);


/* Description: Counts the number of elements in the tree.
 *
 *
 * Params: tree - A pointer to the tree structure.
 *
 * Return Value: number of elements in the tree.
 *
 * Time complexity: O(n).
 *
*/
size_t AVLCount(const AVL_t *tree);


/* Description: A Boolean function designed to indicate presence or lack thereof
 *              of elements in the tree.
 *
 * Params: tree - A pointer to the tree structure.
 *
 * Return Value: Boolean.
 *
*/
int AVLIsEmpty(const AVL_t *tree);


/* Description: A service set to indicate the height of the tree
 *
 * Params: tree - A pointer to the tree structure.
 *
 * Return Value: Height of the tree.
 *
*/
size_t AVLHeight(const AVL_t *tree);


/* Description:apply a function to each element in the given range.
 *
 * Params:
 *         tree    - A pointer to the tree structure.
 *         ActionFunc - A pointer to an action function to be performed on each
 *                      element in the tree.
 *
 *         param - a parameter for the action function.
 *
 * Return Value: Success (0) or return value of ActionFunc on failure.
 *
*/
int AVLForEach(AVL_t *tree, action_func_t ActionFunc, void *param);


/* Description: Finds all occurences of data by key, via is_match function.
                A user supplied doubly linked list will contain all of these 
                instances.
 *
 * Params: tree    - A pointer to the tree structure.
 *         dl_list - user supplied list to contain found cases.
 *         IsMatch - a user function to determine the key by which FindAll 
 *                   determines whether or not a certain element is valid.
 *         arg     - an argument or IsMatch.   
 * Return Value: Boolean.
 *
*/
int AVLFindAll(const AVL_t *tree, is_match_t IsMatch, void *arg, dll_t *list);


/* Description: Removes all occurences of data by key, via is_match function.
                A user supplied doubly linked list will contain all of these 
                instances.
 *
 * Params: tree    - A pointer to the tree structure.
 *         IsMatch - a user function to determine the key by which FindAll 
 *                   determines whether or not a certain element is valid.
 *         arg     - an argument or IsMatch.   
 *
*/
void AVLRemoveAll(AVL_t *tree, is_match_t IsMatch, void *arg);

#endif /* _AVL_H_  */