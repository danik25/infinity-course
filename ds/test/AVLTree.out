/home/xyz/git/dani-benjamin/ds/AVLTree.out